### ISSUE TITLE
RegExp: /u pattern with \u{HexDigits} production with codepoint >0x10FFFF should throw SyntaxError

### ISSUE THREAD

**Issue by dilijev** [Microsoft/ChakraCore#2753]
For Strings:
https://tc39.github.io/ecma262/#prod-UnicodeEscapeSequence
https://tc39.github.io/ecma262/#sec-string-literals-static-semantics-early-errors

> It is a Syntax Error if the MV of HexDigits > 0x10FFFF.

For RegExp (where a string literal is used):
https://tc39.github.io/ecma262/#sec-regexp-pattern-flags

>If pattern is supplied using a StringLiteral, the usual escape sequence substitutions are performed before the String is processed by RegExp. If pattern must contain an escape sequence to be recognized by RegExp, any U+005C (REVERSE SOLIDUS) code points must be escaped within the StringLiteral to prevent them being removed when the contents of the StringLiteral are formed.

Hence the above error applies trivially in this case.

---

For RegExp where a RegExp literal is used:
https://tc39.github.io/ecma262/#sec-patterns-static-semantics-early-errors

> It is a Syntax Error if the MV of HexDigits > 0x10FFFF.

We fail to do this.

---

Test case `regexp-u-110000.js`:

```js
console.log(/\u{110000}/u);
```

Console output:

```
> node --version
v7.7.4

> node -pe "/\u{110000}/u"
[eval]:1
/\u{110000}/u
^

SyntaxError: Invalid regular expression: /\u{110000}/: Invalid unicode escape
    at [eval]:1:1
    at ContextifyScript.Script.runInThisContext (vm.js:23:33)
    at Object.runInThisContext (vm.js:95:38)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (module.js:571:32)
    at Immediate.<anonymous> (bootstrap_node.js:392:29)
    at runCallback (timers.js:666:20)
    at tryOnImmediate (timers.js:639:5)
    at processImmediate [as _immediateCallback] (timers.js:611:5)

> nvs
PATH -= %LOCALAPPDATA%\nvs\node\7.7.4\x64
PATH += %LOCALAPPDATA%\nvs\chakracore-nightly\8.0.0-nightly2017032287ce098018\x64

> node -pe "/\u{110000}/u"
/\u{110000}/u

> node regexp-u-110000.js
/\u{110000}/u

> ch regexp-u-110000.js
/\u{110000}/u
```

Here's a test using UnitTestFramework as well:

`ChakraCore\test\UnifiedRegex\UnicodeEscapeExtended.js`

```js
//-------------------------------------------------------------------------------------------------------
// Copyright (C) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
//-------------------------------------------------------------------------------------------------------

WScript.LoadScriptFile("..\\UnitTestFramework\\UnitTestFramework.js");

let tests = [
    {
        name: "Extended Unicode Escape with out-of-range codepoint should throw",
        body: function () {
            assert.doesNotThrow(function () { /\u{10FFFF}/u });
            assert.throws(function () { /\u{110000}/u }, SyntaxError, "out-of-range codepoint should throw");
        }
    }
];

testRunner.runTests(tests, { verbose: WScript.Arguments[0] != 'summary' });
```

**Comment by EdMaurer** [comment]
What is the impact of this issue?

**Comment by dilijev** [comment]
Very minimal. If a user is requesting a pattern to match codepoint for a character that can never exist by the definition of Unicode codepoints (max value 0x10FFFF), then the pattern will never match any input. The error is supposed to be helpful, but it is not critical.

**Comment by dilijev** [comment]
Cleared assignee so someone else can claim this issue. @sarahdonofrio: please let us know if you want to pick up this issue in the future.

**Comment by dilijev** [comment]
Fixed in #5027
